### PR TITLE
CC: Enable e2e tests for IBM Z secure execution (SE)

### PR DESCRIPTION
**Pull request by BbolroC**
This is to run e2e tests differently based on TEE support for s390x. If the environment is detected, a secure image is built. Otherwise, kernel+rootfs is used for the test.

Fixes: #5429

Signed-off-by: Hyounggyu Choi <Hyounggyu.Choi@ibm.com>

### CODE REVIEW

**Review by hbrueckner — APPROVED** [review]
LGTM! Thanks a lot for your work @BbolroC!

File: integration/kubernetes/confidential/agent_image.bats
Comment (stevenhorsman) [inline integration/kubernetes/confidential/agent_image.bats:125]: What was the reason for moving this test out of interest?

File: integration/kubernetes/confidential/agent_image.bats
Comment (stevenhorsman) [inline integration/kubernetes/confidential/agent_image.bats:184]: I don't like relying on the order of tests for them to pass as it is fragile. I thought that the setup run between tests should have reset the state anyway? If not we need to look at fixing this.

File: lib/common.bash
Comment (stevenhorsman) [inline lib/common.bash:108]: Can we refactor there methods to call the version in kata-containers (or vice versa) rather than duplicating them?

File: integration/kubernetes/confidential/agent_image.bats
Comment (BbolroC) [reply integration/kubernetes/confidential/agent_image.bats:184]: Yeah, as far as I figured it out, [setup()](https://github.com/kata-containers/tests/blob/f07f559f7ea93798458865bbeeca8173e00f6f47/integration/kubernetes/confidential/agent_image.bats#L75-L94) resets a configuration file, but a root file system for an image/initrd.  In [cp_to_guest_img()](https://github.com/kata-containers/tests/blob/f07f559f7ea93798458865bbeeca8173e00f6f47/lib/common.bash#L274-L343), the image/initrd is modified and used for test, but it is not restored during the setup(), which leads to dependency between the tests. I am not sure if this was intentional. But you are right. We could fix this in a separate PR first.

File: integration/kubernetes/confidential/agent_image.bats
Comment (BbolroC) [reply integration/kubernetes/confidential/agent_image.bats:125]: Reasoning is specified below. :wink:

File: lib/common.bash
Comment (BbolroC) [reply lib/common.bash:108]: yep, I will refactor them. Thanks for pointing it out. :+1:

File: integration/kubernetes/confidential/agent_image.bats
Comment (stevenhorsman) [reply integration/kubernetes/confidential/agent_image.bats:184]: I guess that leads me to the question of how they worked before you moved them? Is the difference due to the agent-config.toml now being added in the rootfs.sh rather than us doing that during the test time?

Comment (stevenhorsman) [reply integration/kubernetes/confidential/agent_image.bats:184]: Interestingly I've just spotted that the crictl test have the follow in their teardown:
https://github.com/kata-containers/tests/blob/f07f559f7ea93798458865bbeeca8173e00f6f47/integration/containerd/confidential/agent_api.bats#L79-L81
so maybe we can just add this to the Kubernetes teardown as well to resolve the problem.
@wainersm - can you recall if there is a difference in the way that the guest image setup works for crictl and Kubernetes tests?

**Comment by BbolroC** [comment]
@stevenhorsman all tests are passed with the original order after having `build_se_image()` called when kernel parameters are configured during `setup()` and `teardown()`. I overlooked an update of the secure image with a new parmfile (including kernel parameters) and only focused on change of rootfs. This explains why we experienced such a weird phenomenon (2 tests cases we looked into only modify kernel parameters, but rootfs. So we were not able to see what we had expected.)

File: integration/confidential/lib.sh
Comment (stevenhorsman) [inline integration/confidential/lib.sh:58]: I'm a bit torn on the check being in `switch_measured_rootfs_verity_scheme`. On the onehand having it in just a single place is good, but I don't know if it would be obvious to anyone adding tests to `agent_image.bats`?

I think it is fine as is, just thinking out loud a bit!

**Review by stevenhorsman — APPROVED** [review]
I have a few minor comments/questions, but generally I think it looks good. I just wanted to check what the next steps are:
My understanding is that we don't have a public SE-enabled machine yet, if this is correct, could you add the results of your local run to the PR just so we can see the output? Also is the plan for this to be run behind the IBM firewall but triggered on PRs (like the tdx tests), or just manually ad-hoc until we have a box we can add to jenkins?

Thanks!

File: integration/confidential/lib.sh
Comment (BbolroC) [reply integration/confidential/lib.sh:58]: I agree with your thought. (I was thinking it actually :wink:) It would be something to be relocated based on others' feedback. Let's see what others think about this placement.

File: lib/common.bash
Comment (BbolroC) [reply lib/common.bash:15]: I am totally for your opinion. I will refactor the code as advised. Thanks a lot!

File: lib/common.bash
Comment (BbolroC) [reply lib/common.bash:396]: yes, as far as I know, that is the only way to get the change of rootfs reflected on the se image. But I would like to listen to @hbrueckner 's opinion as well.

**Comment by hbrueckner** [comment]
Hi @stevenhorsman 

> I have a few minor comments/questions, but generally I think it looks good. I just wanted to check what the next steps are: My understanding is that we don't have a public SE-enabled machine yet, if this is correct, could you add the results of your local run to the PR just so we can see the output? 

Providing a public SE-enabled machine is still work in progress. Until it will be available, would it be an option to share results from internal testing to move forward and get it merged? Wdyt?

**Comment by stevenhorsman** [comment]
> Providing a public SE-enabled machine is still work in progress. Until it will be available, would it be an option to share results from internal testing to move forward and get it merged? Wdyt?

As long as the changes don't have any negative impact on the other platforms (so we have green tests as usual) that sounds okay to me

File: lib/common.bash
Comment (BbolroC) [reply lib/common.bash:120]: Yep, I will make them as little different as possible. Thanks.

**Review by GabyCT — APPROVED** [review]

**Comment by BbolroC** [comment]
@fidencio @stevenhorsman I've created a Jenkins job which fetches and dumps the log of the internal test. The test runs on a daily basis and the job will be updated accordingly. Here is a result of the first run:
http://jenkins.katacontainers.io/job/kata-containers-CCv0-ubuntu-20.04-s390x-SE-daily/1/consoleText

**Comment by stevenhorsman** [comment]
> @fidencio @stevenhorsman I've created a Jenkins job which fetches and dumps log of the internal test. The test runs on a daily basis and the job will be triggered accordingly. Here is a result of the first run: http://jenkins.katacontainers.io/job/kata-containers-CCv0-ubuntu-20.04-s390x-SE-daily/1/consoleText

This is great to see. A couple of questions:
- I assume it is running the latest 'CCv0' branch each night?
- If there is a failure on the internal tests, with the results be marked red to reflect that?
Thanks!

**Comment by stevenhorsman** [comment]
> Yeah, it is totally viable, but there is an internal issue (flaky) on the test `Test can pull an encrypted image inside the guest with decryption key` regarding:
> 
> ```
> failed to pull manifest Registry error: url https://registry-1.docker.io/v2/katadocker/ccv0-ssh/manifests/s390x,
> envelope: OCI API errors: [OCI API error: You have reached your pull rate limit. You may increase the limit by authenticating
> ```
> 
> It looks there is a company policy for the `docker.io` plan, which it is a bit challenging for me to resolve. It would be great if we could switch the registry to `quay.io`.

I know Intel have hit this as well and I think also AMD, so it's something that is being discussed. There is an issue with quay.io and skopeo having a difference of opinion on schema version IIRC, so I know in the past when Jakob and I tried to push the encrypted image there it didn't work and I think Tobin/Ryan found the same more recently. I think Fabiano might have reached out to Hendrik to see if use using ICR would be possible, but I don't know if that went anywhere.

I think when the KBS URI changes that Ding has proposed go in happen then all the existing images we use will be invalidated, so I hope at that point we can ensure we create proper multiarch images into a container registry that is reliably, but we'll see.

This problem shouldn't block this PR anyway.

**Comment by BbolroC** [comment]
Why don't we merge this unless the test failures for tdx have something to do with this change?

**Comment by stevenhorsman** [comment]
I'm assuming that this has a dependency on https://github.com/kata-containers/kata-containers/pull/6207, so we need to get that merged first?

**Comment by fidencio** [comment]
Also network related issues here. :-/

**Comment by fidencio** [comment]
QEMU TDX CI has passed but didn't report back: http://10.112.240.208:8080/job/tests-CCv0-qemu-TDX-CRI_CONTAINERD_K8S-PR/387/

**Comment by fidencio** [comment]
Cloud Hypervisor CI has passed but didn't report back: http://10.112.240.208:8080/job/tests-CCv0-cloud-hypervisor-TDX-CRI_CONTAINERD_K8S-PR/384/

**Pull request merged**
